### ISSUE TITLE
Scrub Running Indefinitely; Several Kernel Errors, Including Panic at arc.c:1544:arc_buf_type()

### ISSUE THREAD

**Issue by Azratosh** [openzfs/zfs#12929]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Debian
Distribution Version	| 11 (bullseye)
Kernel Version	| [5.10.0-9-amd64](https://packages.debian.org/bullseye/linux-image-5.10.0-9-amd64)
Architecture	| AMD64
OpenZFS Version	| zfs-2.1.2-1~bpo11+1

### Describe the problem you're observing
My zpool would randomly hang while running a scrub, which eventually caused my entire server to freeze. After logging `zpool status` every 60 seconds, I found out that the scrub would just hang after a while, causing it to run indefinitely without making progress. (Forcibly) rebooting would cause the scrub progress to reset to a certain point.

An interesting thing I've observed, is that my pool remains almost fully functional -- I can browse my files on it, which includes mostly large videos, but also smaller files that are synced on my Nextcloud. However, I found out that running `ls -R /pool-00` hangs at a certain directory in my video archive, always, seemingly not being able to descend into it. The same held true for another sub-directory, one where some Nextcloud data is stored.
**The sub-directories were accessible again after a reboot.**

The issues only started to arise after updating to 2.1.2; I believe the previous version I was running was 2.0.4.

### Describe how to reproduce the problem
Hard to reproduce ad hoc; seems to happen during normal usage and scrubbing.

### Include any warning/errors/backtraces from the system logs

```
$ date; sudo zpool status
Sun  2 Jan 15:21:22 CET 2022
  pool: pool-00
 state: ONLINE
status: One or more devices has experienced an unrecoverable error.  An
        attempt was made to correct the error.  Applications are unaffected.
action: Determine if the device needs to be replaced, and clear the errors
        using 'zpool clear' or replace the device with 'zpool replace'.
   see: https://openzfs.github.io/openzfs-docs/msg/ZFS-8000-9P
  scan: scrub in progress since Wed Dec 29 23:21:38 2021
        14.5T scanned at 33.5M/s, 11.0T issued at 16.1M/s, 15.0T total
        1.79M repaired, 73.21% done, 3 days 00:39:02 to go
config:

        NAME                                 STATE     READ WRITE CKSUM
        pool-00                              ONLINE       0     0     0
          raidz1-0                           ONLINE       0     0     0
            ata-ST8000VN004-2M2101_WKD136MM  ONLINE       0     0     5  (repairing)
            ata-ST8000VN004-2M2101_WKD15F18  ONLINE       0     0     4  (repairing)
            ata-ST8000VN004-2M2101_WRD09RAH  ONLINE       0     0     5  (repairing)
            ata-ST8000VN004-2M2101_WRD0ABM6  ONLINE       0     0     5  (repairing)
        logs
          sde1                               ONLINE       0     0     0
        cache
          sde2                               ONLINE       0     0     0

errors: No known data errors
```
Shutting down hanged, so I had to kill my server via the power button. 
**After booting again:**
```
$ date; sudo zpool status
Sun  2 Jan 15:28:16 CET 2022
  pool: pool-00
 state: ONLINE
  scan: scrub in progress since Wed Dec 29 23:21:38 2021
        10.7T scanned at 26.8G/s, 7.69T issued at 137M/s, 15.0T total
        1012K repaired, 51.39% done, 15:25:38 to go
config:

        NAME                                 STATE     READ WRITE CKSUM
        pool-00                              ONLINE       0     0     0
          raidz1-0                           ONLINE       0     0     0
            ata-ST8000VN004-2M2101_WKD136MM  ONLINE       0     0     0
            ata-ST8000VN004-2M2101_WKD15F18  ONLINE       0     0     0
            ata-ST8000VN004-2M2101_WRD09RAH  ONLINE       0     0     0
            ata-ST8000VN004-2M2101_WRD0ABM6  ONLINE       0     0     0
        logs
          sde1                               ONLINE       0     0     0
        cache
          sde2                               ONLINE       0     0     0

errors: No known data errors
```

I have attached a bunch of syslog snippets (all I could find, really) and a log of SMART data of all drives currently used by my zpool. The most notable one is `01-arc-evict.txt`, where a Kernel panic occured at `at arc.c:1544:arc_buf_type()`.

[01-arc-evict.txt](https://github.com/openzfs/zfs/files/7799221/01-arc-evict.txt)
[02-soft-lockup.txt](https://github.com/openzfs/zfs/files/7799222/02-soft-lockup.txt)
[03-invalid-opcode.txt](https://github.com/openzfs/zfs/files/7799223/03-invalid-opcode.txt)
[04-general-protection-fault.txt](https://github.com/openzfs/zfs/files/7799224/04-general-protection-fault.txt)
[05-general-protection-fault.txt](https://github.com/openzfs/zfs/files/7799225/05-general-protection-fault.txt)
[06-invalid-raidz-configuration.txt](https://github.com/openzfs/zfs/files/7799226/06-invalid-raidz-configuration.txt)
[07-list-del-corruption.txt](https://github.com/openzfs/zfs/files/7799227/07-list-del-corruption.txt)
[08-general-protection-fault.txt](https://github.com/openzfs/zfs/files/7799228/08-general-protection-fault.txt)
[disks_after_long_2022-01-02.txt](https://github.com/openzfs/zfs/files/7799229/disks_after_long_2022-01-02.txt)

**Comment by Azratosh** [comment]
And some more logs, after another hangup.

Output of `dmesg`:
[dmesg.log](https://github.com/openzfs/zfs/files/7799555/dmesg.log)
Output of `zpool events -v`:
[events.log](https://github.com/openzfs/zfs/files/7799556/events.log)

Appreciate any help!

**Comment by rincebrain** [comment]
Debian shipped 2.0.3 with patches for a long time, then jumped to 2.0.6, then 2.1.x, I believe.

That said, that's very exciting. What are the properties on the pool/dataset(s) you're using, and what hardware are you running it on? Have you tried leaving memtest running for a little while to eliminate that as a possible option? (I don't think this is especially likely to be a hardware bug, but when you hit error codepaths I've never seen anyone else hit, I ask.)

When ZFS hits a fundamental assertion failure, by default it tells the thread to go hang out dead forever (the other option is triggering a hard panic(), if you like - spl_panic_on_halt is I believe the tunable for that), so anything that tried to touch shared resources that thread was using would hang. And since you were in the read cache code when that happened...

**Comment by Azratosh** [comment]
I will postpone sending you my hardware info, because... well, see for yourself:
![image](https://user-images.githubusercontent.com/26200153/147986169-44ca5cce-91d7-453f-bce5-2c715dfe12f1.jpeg)

Turns out my new RAM is borked. This also explains why my zpool suddenly had a bunch of checksum errors. I was doing weekly scrubs and automatically let it mail me the output of `zpool status`. Sigh. Guess I'll have to RMA it. I just hope none of my data is corrupted beyond repair.

I believe we can consider this issue as closed now. Thanks a bunch for pointing me towards memtesting!
